rllib/examples/envs/env_rendering_and_recording.py

```diff
@@ -1,135 +1,274 @@
-    }
-
-    def __init__(self, config):
-        self.end_pos = config.get("corridor_length", 10)
-        self.max_steps = config.get("max_steps", 100)
-        self.cur_pos = 0
-        self.steps = 0
-        self.action_space = Discrete(2)
-        self.observation_space = Box(0.0, 999.0, shape=(1,), dtype=np.float32)
-
-    def reset(self, *, seed=None, options=None):
-        self.cur_pos = 0.0
-        self.steps = 0
-        return [self.cur_pos], {}
-
-    def step(self, action):
-        self.steps += 1
-        assert action in [0, 1], action
-        if action == 0 and self.cur_pos > 0:
-            self.cur_pos -= 1.0
-        elif action == 1:
-            self.cur_pos += 1.0
-        truncated = self.steps >= self.max_steps
-        done = self.cur_pos >= self.end_pos or truncated
-        return [self.cur_pos], 10.0 if done else -0.1, done, truncated, {}
-
-    def render(self, mode="rgb"):
-        """Implements rendering logic for this env (given current state).
-
-        You can either return an RGB image:
-        np.array([height, width, 3], dtype=np.uint8) or take care of
-        rendering in a window yourself here (return True then).
-        For RLlib, though, only mode=rgb (returning an image) is needed,
-        even when "render_env" is True in the RLlib config.
-
-        Args:
-            mode: One of "rgb", "human", or "ascii". See gym.Env for
-                more information.
-
-        Returns:
-            Union[np.ndarray, bool]: An image to render or True (if rendering
-                is handled entirely in here).
+
+
+class EnvRenderCallback(DefaultCallbacks):
+    """A custom callback to render the environment.
+
+    This can be used to create videos of the episodes for some or all EnvRunners
+    and some or all env indices (in a vectorized env). These videos can then
+    be sent to e.g. WandB as shown in this example script here.
+
+    We override the `on_episode_step` method to create a single ts render image
+    and temporarily store it in the Episode object.
+    """
+
+    def __init__(self):
+        super().__init__()
+        # Per sample round (on this EnvRunner), we want to only log the best- and
+        # worst performing episode's videos in the custom metrics. Otherwise, too much
+        # data would be sent to WandB.
+        self.best_episode_and_return = (None, float("-inf"))
+        self.worst_episode_and_return = (None, float("inf"))
+
+    def on_episode_step(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """On each env.step(), we add the render image to our Episode instance.
+
+        Note that this would work with MultiAgentEpisodes as well.
+        """
+        # If we have a vector env, only render the sub-env at index 0.
+        if isinstance(env.unwrapped, gym.vector.VectorEnv):
+            image = env.envs[0].render()
```

Comment:
Dumb question: Imagine a user needs to return a couple of images in array form. Does she return then a `np.ndarray(shape(K, H, W, 3))` with `K` the number of images? 

And if we have a vector environment, how does this look like in Tensorboard?

Reply:
I like however that we use the standard API of `gymnasium` here and the user just needs to override this.

Reply:
Yes, me, too. This was a total mess in the old RLlib, with the `render_env` option.

Reply:
> Dumb question: Imagine a user needs to return a couple of images in array form. Does she return then a `np.ndarray(shape(K, H, W, 3))` with `K` the number of images?

Correct (I think this is even from your own PR a while back, where you enabled the WandB logger to actually log images and videos):
* shape=3D -> single image e.g. [c, h, w]
* shape=4D -> n images (all of the same size) e.g. [N, c, h, w]
* shape=5D -> 1 video with shape [L, c, h, w]

c=channels, L=length of video, w=width, h=height

> And if we have a vector environment, how does this look like in Tensorboard?